draft-ppm-protocol.md

```diff
@@ -647,147 +676,222 @@ be cryptographically protected as described in {{helper-state}}.
-                     "pda input share" || task_id || server_role)
-input_share = context.Open(time || nonce || extensions, helper_share)
+                     "pda input share" || task_id || 0x01)
+input_share = context.Open(time || nonce || extensions,
+                           helper_share.payload)
 ~~~
 
-where `sk` is the HPKE secret key, `task_id` is `AggregateReq.task_id` and
-`server_role` is the role of the server (`0x01` for the leader and `0x00` for
-the helper). `time`, `nonce` and `extensions` are obtained from the
-corresponding fields in `AggregateSubReq`. If decryption fails, then the
-sub-response consists of a "decryption error" alert. [See issue#57.] Otherwise,
-the helper handles the request for its plaintext input share `input_share` and
-updates its state as specified by the PPM protocol.
+where `task_id` is `VerifyStartReq.task_id` and `time`, `nonce`, and
+`extensiosn` are obtained from the corresponding fields in `VerifyStartSubReq`.
+If decryption fails, then the sub-response consists of a "decryption error"
+alert. [See issue#57.]
+
+Otherwise, the helper runs the VDAF verify-start algorithm to compute its local
+VDAF state and the verification message:
+
+~~~
+(state, VerifySubResp.verify_message) = vdaf_start(0x01,
+          verify_param, output_param, time || nonce, input_share)
+~~~
 
 After processing all of the sub-requests, the helper encrypts its updated state
-and constructs its response to the aggregate request.
+and constructs its response to the request.
+
+### Verify-Next Request
+
+For VDAFs requiring more than one round of communication, subsequent rounds are
+handled by using a *verify-next request*. the leader sends a POST request to
+`[aggregator]/aggregate_next` with the following message:
+
+~~~
+struct {
+  TaskID task_id;
+  opaque helper_state<0..2^16>;
+  VerifyNextSubReq seq<1..2^24-1>;
+} VerifyNextReq;
+~~~
+
+* `task_id` is the same as task ID as the verify-start request.
+* `helper_state` is the opaque state sent by the helper in response to the
+  previous verify request.
+* `seq` is the sequence of sub-requests. It has the same number of sub-requests
+  as the previous request.
+
+Each sub-request has the following structure:
+
+~~~
+struct {
+  Time time;
+  uint64 nonce;
+  opaque verify_message<1..2^16-1>;
+} VerifyNextSubReq;
+~~~
+
+Fields `time` and `nonce` have the same value as the corresponding sub-request
+in the verify-start request. The helper's response consists of a `VerifyResp`
+message constructed as described above.
+
+Both aggregators compute verification message using the VDAF verify-next
+algorithm as
+
+~~~
+(state, AggregateSubReq.verify_message) = vdaf_next(
+          state, inbound_message)
+~~~
+
+where `inbound_message` is the verification message sent by the peer in response
+to the previous request and `state` is the aggregator's VDAF state.
+
+### Recovering the Output Share
+
+Once all rounds of the VDAF have been completed for a given report, each
```

Comment:
The helper has nowhere but the `helper_state` blob to keep track of how many rounds have been executed. How does it know whether to execute `vdaf_next` or `vdaf_finish` for any `VerifyNextSubReq`? Perhaps we are missing something like `vdaf_is_finished(state) -> int` (but with a good name instead of that bad one)?

Reply:
The VDAF fixes the number of rounds. The first round uses vdaf_start and subsequent rounds use vdaf_next.

Reply:
I buy that the number of rounds is a constant determined by the VDAF. But how does helper know what round it's on when it receives a `VerifyNextReq`? Implicitly a helper would have to keep a count of executed rounds in its `helper_state`, or we could require the VDAF to encode that into the `state` that gets passed into `vdaf_next`.

Reply:
I'd also be fine with adding a "round_number" field to the request body.

Reply:
That could work, but then we have to consider what if any risks there are to allowing the leader to control the round number. If we put it in helper-controlled state, then it's protected by the encryption we already require from helper implementations.

Reply:
Good point! Either way the helper is going to have to keep track of its state, and thus what message to expect next. I think we should make no changes here and just make it clear that the helper has to keep track of this.